src/devices/sound/hc55516.cpp

```diff
@@ -21,242 +31,561 @@
+	//stub, don't even bother registering for savestates, this should NEVER be called
 }
 
 //-------------------------------------------------
 //  device_reset - device-specific reset
 //-------------------------------------------------
 
-void hc55516_device::device_reset()
+void cvsd_device::device_reset()
 {
 	m_last_clock_state = 0;
 }
 
-DEFINE_DEVICE_TYPE(MC3417, mc3417_device, "mc3417", "MC3417")
+//-------------------------------------------------
+//  device_clock_changed - device-specific samplerate change
+//-------------------------------------------------
+/*void cvsd_device::device_clock_changed()
+{
+	// do nothing.
+	//m_stream->set_sample_rate(clock());
+}*/
 
-mc3417_device::mc3417_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
-	: hc55516_device(mconfig, MC3417, tag, owner, clock)
+READ_LINE_MEMBER( cvsd_device::clock_r )
 {
+	// prevent debugger from changing the internal state
+	if (!machine().side_effects_disabled())
+		m_stream->update(); /* bring up to date first */
+	return clock_state_r();
 }
 
-//-------------------------------------------------
-//  device_start - device-specific startup
-//-------------------------------------------------
+WRITE_LINE_MEMBER( cvsd_device::mclock_w )
+{
+	clock_w(state);
+}
 
-void mc3417_device::device_start()
+WRITE_LINE_MEMBER( cvsd_device::digin_w )
 {
-	start_common(0x07, false);
+	digit_w(state);
 }
 
+// the following encode related functions don't do anything yet, don't call them.
+void cvsd_device::audio_in_w(int16_t data)
+{
+	assert(0);
+}
 
-DEFINE_DEVICE_TYPE(MC3418, mc3418_device, "mc3418", "MC3418")
+WRITE_LINE_MEMBER( cvsd_device::dec_encq_w )
+{
+	assert(0);
+}
 
-mc3418_device::mc3418_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
-	: hc55516_device(mconfig, MC3418, tag, owner, clock)
+READ_LINE_MEMBER( cvsd_device::digout_r )
 {
+	return 0;
 }
 
-//-------------------------------------------------
-//  device_start - device-specific startup
-//-------------------------------------------------
+// default and stub implementations
 
-void mc3418_device::device_start()
+inline bool cvsd_device::is_external_oscillator()
 {
-	start_common(0x0f, false);
+	return clock() != 0;
 }
 
+inline bool cvsd_device::is_clock_changed(bool clock_state)
+{
+	return ((!m_last_clock_state && clock_state) ||
+			(m_last_clock_state && !clock_state));
+}
 
-void hc55516_device::start_common(uint8_t _shiftreg_mask, int _active_clock_hi)
+inline bool cvsd_device::is_active_clock_transition(bool clock_state)
 {
-	/* compute the fixed charge, decay, and leak time constants */
-	m_charge = pow(exp(-1.0), 1.0 / (FILTER_CHARGE_TC * 16000.0));
-	m_decay = pow(exp(-1.0), 1.0 / (FILTER_DECAY_TC * 16000.0));
-	m_leak = pow(exp(-1.0), 1.0 / (INTEGRATOR_LEAK_TC * 16000.0));
+	return (( m_active_clock_edge && !m_last_clock_state &&  clock_state) ||
+			(!m_active_clock_edge &&  m_last_clock_state && !clock_state));
```

Comment:
This simplifies to `(bool(clock_state) != m_last_clock_state) && (bool(clock_state) == m_active_clock_edge)`

Reply:
Fixed as suggested.